RONDB-860: Ensure that pk-read is part of relativeUrl

mronstro opened the pull request.

mikaelronstrom committed:
RONDB-860: Ensure that pk-read is part of relativeUrl

KernelMaker commented on storage/ndb/rest-server2/server/src/json_parser.cpp:
```diff
@@ -1457,17 +1458,26 @@ RS_Status extract_db_and_table(const std::string_view &relativeUrl,
+      std::string(rdrsErrorMessage(ERROR_INVALID_RELATIVE_URL))).status;
   }
-  return CRS_Status::SUCCESS.status;
+  if (request_type.length() == strlen(PKREAD) &&
```
This if statement doesn't have an effect

mikaelronstrom committed:
RONDB-860: Handled edge cases of relativeUrl better

mikaelronstrom committed:
RONDB-860: Use one transaction per operation for RDRS batch requests

When issuing a batch request, it should be treated as a set of pk-read requests
in RDRS. This means that we can use one transaction per operation.

Using one transaction per operation means that we can send the operation to the
nodes where the primary replica resides directly which will cause load balancing
and parallel reads to start already from the NDB API. With a single transaction
all requests have to go through one node and one TC thread.

Here we scale on number of nodes and on number of TC threads per node.

The overhead of a transaction is almost neglible, so no issue on that front.

mikaelronstrom committed:
RONDB-860 RONDB-791: Fixes for batchpkread

Handling of relative_url in both test programs and RDRS2 was faulty.
The acceptable format is the following ones:
db/table/pk-read
/db/table/pk-read
db/table/pk-read/
/db/table/pk-read/

Fixed some problems related to handling one transaction per operation.
First Ndb object need to have space for a large number of transactions
executed in parallel (set to max of 1024).
Second the return from sendPollNdb is the number of transactions ready.

All other variants should generate an error.
This required changes in test programs and
in the function extract_db_and_table.

RONDB-791:
----------

404 is generated when a row isn't found
404 is also generated when a table used in the request doesn't exist
404 is now also generated when a column referred to doesn't exist in the given table

Minor fixes of format, avoid using more than 80 chars on a line.

mikaelronstrom committed:
RONDB-862: Fixed handling of errno

In the case of strtol and strtoll errno is not set before returning
successfully. This means that an old errno value can look like a failed
call to strtol(l). To avoid this a lot of setting errno = 0 was introduced,
in particular at the start of most threads and programs.

In addition some bug fixes around handling return from strtol(l) was
done.

KernelMaker commented:
**Fixing relative-url Issues:**
Two issues still remain:

1. The relative-url parser seems to only remove one extra / at the end. If the URL has more than one trailing /, it treats it as a failure.
2. In batch pk-read, each operation's result should be processed separately. However, with this patch, if even one operation contains an invalid relative-url, the entire batch pk-read request fails.

**Fixing errno Handling:**
I'm not sure if introducing such a large modification (60+ files) is necessary for handling errno. The patch explicitly resets errno = 0 at the beginning of each thread, even for threads that don’t interact with errno at all. This seems excessive, and I’m unsure if it could cause unintended issues in the future.

Suggested Approach:
We should keep the fix as simple as possible by modifying only the places where strtol is used. A more reasonable solution would be to check for errno based on the second parameter of strtol:

```
errno = 0;
char *end;
long value = strtol(str, &end, 10);
if (end > str && *end == '\0') {
    // Success
} else {
    // Handle error
    errno;
}
```
BTW, I also double-checked how MySQL and Redis handle strtol(). In most cases, they simply use it without checking, and only in a few places do they implement checks similar to the approach above.

mikaelronstrom committed:
RONDB-862: Focus on ensuring that using errno is safe
